[gue]: adding gue decapsulation feature

tehnerd opened the pull request:
we want to be on for GUE w/ IPIP. in this diff i'm adding support
for GUE based decapsulation and making sure that features, like src
based routing (when we send packet to backend based on src ip address.
used in applications like stable/stabilized anycast
(https://landing.google.com/sre/workbook/chapters/managing-load/#stabilized-anycast))

there some added verbosity in defines dependency, but this is mostly for
not to break existing build (e.g. if someone is building katran prior
this diff w/ LPM_SRC_LOOKUP, he assumes that INLINE_DECAP would be
enabled as well w/o explicitly configureing INLINE_DECAP_IPIP).
happy to remove this complexity if you think that it acceptable to break
this builds.

also made sure that we could do both IPIP and GUE decaps at the same
time (e.g. for migration purpose. when katran could receive both GUE and
IPIP packets from remote side and decapsulate em for subsequent lookups;
to make migration to gue possible)

current size of bpf program when all the features are enabled
( ./build_bpf_modules_opensource.sh "-DICMP_TOOBIG_GENERATION
-DGUE_ENCAP -DINLINE_DECAP_GUE -DLPM_SRC_LOOKUP -DINLINE_DECAP_IPIP
-DKATRAN_INTROSPECTION")
is 2739 instructions

Tested-By: added new optional unittests for GUE

udippant commented on katran/lib/bpf/balancer_kern.c:
```diff
@@ -236,7 +236,35 @@ static inline int process_l3_headers(struct packet_description *pckt,
+
+#ifdef INLINE_DECAP_IPIP
 __attribute__((__always_inline__))
 static inline int process_encaped_pckt(void **data, void **data_end,
```
nit: rename this to process_encapped_ipip_pckt?

udippant commented on katran/lib/bpf/balancer_consts.h:
```diff
@@ -25,6 +25,11 @@
+
+// GUE varian 1 using first four bits of inner packet as a pseudo header
+// we are using last two of this four bits to distinct v4 vs v6. see RFC for more details
+#define GUEV1_IPV6MASK 0x30
```
Also, so that we don't forget - this mask https://tools.ietf.org/html/draft-ietf-intarea-gue-05?fbclid=IwAR39fGS_slDK4PFeg1uTTcE8dGJ79AhgEJjnkQ1xm7U0IISNXYMX-l5oQ2I#section-4

tehnerd replied:
actually mask is corrent. we want to know is it v6 or v4. for v4 first 4 bits are:
0100
for v6:
0110

our mask is 0011 (+0000; but we only care about first 4 bits)

v4 & mask == 0
v6 & mask != 0. so it is working as intended

tehnerd commented on katran/lib/bpf/pckt_encap.h:
```diff
@@ -356,6 +356,43 @@ static inline bool gue_encap_v6(struct xdp_md *xdp, struct ctl_value *cval,
+}
+
+__attribute__((__always_inline__)) static inline bool
+gue_decap_v6(struct xdp_md* xdp, void** data, void** data_end, bool inner_v4) {
```
for decap it is "inner_v4" for IPIP as well. as we need special handling only if inner is, well, v4 when outer is v6

tehnerd commented:
i've force pushed new version w/ 
1. fixed typos 
2. renamed process_encaped_packets to encaped_ipip_packets

i did not change mask (as it is correct), nor changed "inner_v4" as it is on par w/ ipip decapsulation and actually make more sense imo in that case
